sklearn/inspection/_partial_dependence.py

```diff
@@ -94,46 +99,81 @@ def _grid_from_X(X, percentiles, is_categorical, grid_resolution):
     if grid_resolution <= 1:
         raise ValueError("'grid_resolution' must be strictly greater than 1.")
 
+    custom_values = {k: np.asarray(v) for k, v in custom_values.items()}
```

Comment:
I think the issue with the default resulting type in `cartesian` comes from using `np.asarray` on a list of strings creates a "string array". I think we should make sure to have object dtypes here when creating `custom_values`:

```suggestion
    def _convert_custom_values(values):
        # Convert custom types such that object types are always used for string arrays
        dtype = object if any(isinstance(v, str) for v in values) else None
        return np.asarray(values, dtype=dtype)

    custom_values = {k: _convert_custom_values(v) for k, v in custom_values.items()}
```